dataflow/src/main/java/org/checkerframework/dataflow/expression/JavaExpression.java

```diff
@@ -660,4 +678,157 @@ public static JavaExpression getPseudoReceiver(TreePath path, TypeMirror enclosi
+        List<JavaExpression> parametersJe = new ArrayList<>(methodTree.getParameters().size());
+        for (VariableTree param : methodTree.getParameters()) {
+            parametersJe.add(new LocalVariable(TreeUtils.elementFromDeclaration(param)));
+        }
+        return ViewpointAdaptJavaExpression.viewpointAdapt(this, parametersJe);
+    }
+
+    /**
+     * Viewpoint-adapts {@code this} to the {@code methodInvocationTree}.
+     *
+     * @param methodInvocationTree method invocation
+     * @return viewpoint-adapted version of this
+     */
+    public final JavaExpression atMethodInvocation(MethodInvocationTree methodInvocationTree) {
+        List<JavaExpression> argumentsJe =
+                argumentTreesToJavaExpressions(
+                        methodInvocationTree, methodInvocationTree.getArguments());
+
+        JavaExpression receiverJe = JavaExpression.getReceiver(methodInvocationTree);
+        return ViewpointAdaptJavaExpression.viewpointAdapt(this, receiverJe, argumentsJe);
+    }
+
+    /**
+     * Viewpoint-adapts {@code this} to the {@code invocationNode}.
+     *
+     * @param invocationNode method invocation
+     * @return viewpoint-adapted version of this
+     */
+    public final JavaExpression atMethodInvocation(MethodInvocationNode invocationNode) {
+        List<JavaExpression> argumentsJe = new ArrayList<>(invocationNode.getArguments().size());
+        for (Node argTree : invocationNode.getArguments()) {
+            argumentsJe.add(JavaExpression.fromNode(argTree));
+        }
+
+        Node receiver = invocationNode.getTarget().getReceiver();
+        JavaExpression receiverJe = JavaExpression.fromNode(receiver);
+        return ViewpointAdaptJavaExpression.viewpointAdapt(this, receiverJe, argumentsJe);
+    }
+
+    /**
+     * Viewpoint-adapts {@code this} to the {@code newClassTree}.
+     *
+     * @param newClassTree constructor invocation
+     * @return viewpoint-adapted version of this
+     */
+    public JavaExpression atConstructorInvocation(NewClassTree newClassTree) {
+        List<JavaExpression> argumentsJe =
+                argumentTreesToJavaExpressions(newClassTree, newClassTree.getArguments());
+
+        JavaExpression receiverJe = JavaExpression.getReceiver(newClassTree);
+        return ViewpointAdaptJavaExpression.viewpointAdapt(this, receiverJe, argumentsJe);
+    }
+
+    /**
+     * Converts method or constructor arguments from Trees to JavaExpressions, accounting for
+     * varargs.
+     *
+     * @param tree invocation of the method or constructor
+     * @param argTrees the arguments to the method or constructor; subexpressions of {@code tree}
+     * @return the arguments, as JavaExpressions
+     */
+    private static List<JavaExpression> argumentTreesToJavaExpressions(
+            ExpressionTree tree, List<? extends ExpressionTree> argTrees) {
+
+        if (tree.getKind() == Kind.METHOD_INVOCATION) {
```

Comment:
How can this be false?  The specification in the Javadoc says `tree` is an invocation.

Reply:
Tree can be a NewClassTree, which is a constructor invocation.

Reply:
The Javadoc says `tree` is a method or constructor invocation.  If `tree` is a constructor invocation, then the tree kind is `NEW_CLASS`.